Allow hiding the diagnostic severity in `ruff_db`

ntBre opened the pull request:
## Summary

This PR is a spin-off from https://github.com/astral-sh/ruff/pull/19415. It enables replacing the severity and lint name in a ty-style diagnostic:

```
error[unused-import]: `os` imported but unused
```

with the noqa code and optional fix availability icon for a Ruff diagnostic:

```
F401 [*] `os` imported but unused
F821 Undefined name `a`
```

or nothing at all for a Ruff syntax error:

```
SyntaxError: Expected one or more symbol names after import
```

Ruff adds the `SyntaxError` prefix to these messages manually.

Initially (d912458), I just passed a `hide_severity` flag through a bunch of calls to get it into `annotate-snippets`, but after looking at it again today, I think reusing the `None` severity/level gave a nicer result. As I note in a lengthy code comment, I think all of this code should be temporary and reverted when Ruff gets real severities, so hopefully it's okay if it feels a little hacky.

I think the main visible downside of this approach is that we can't style the asterisk in the fix availabilty icon in cyan, as in Ruff's current output. It's part of the message in this PR and any styling gets overwritten in `annotate-snippets`.

<img width="400" height="342" alt="image" src="https://github.com/user-attachments/assets/57542ec9-a81c-4a01-91c7-bd6d7ec99f99" />

Hmm, I guess reusing `Level::None` also means the `F401` isn't red anymore. Maybe my initial approach was better after all. In any case, the rest of the PR should be basically the same, it just depends how we want to toggle the severity.

## Test Plan

New `ruff_db` tests. These snapshots should be compared to the two tests just above them (`hide_severity_output` vs `output` and `hide_severity_syntax_errors` against `syntax_errors`).

ntBre commented:
I thought it would be a bit repetitive to render `invalid-syntax` in front of `SyntaxError`. I think it would disrupt the concise output to strip `SyntaxError: ` from the front of the actual message contents (unless we literally `strip_prefix` here, which I did consider in one early draft), but I agree with you otherwise.

MichaReiser commented:
Can't we remove the `SyntaxError:` prefix in whatever place we construct those diagnostics

MichaReiser commented:
> Edit: after trying this, we'd actually have to move the prefix into every output format except full to avoid changing their snapshots. (And except gitlab, which already uses strip_prefix)

I'm sort of fine if they change. They should all contain the lint id and the message, so that the prefix seems unnecessary

MichaReiser reviewed (commented):
I now took a closer look. I'd prefer to avoid the invalid syntax special casing. An empty id feels hacky. I'd simply make invalid syntax an error like every other one (except that we use the `diagnostic.id` instead of the `secondary_code` as we should for all errors without a secondary code)

ntBre commented:
Okay I think I've implemented what we discussed here and on Discord for the `full` and `concise` formats. In short, we just replaced the manual `SyntaxError:` prefix with `invalid-syntax:`, which reuses the `LintId`. All of the snapshots in b8aa43f515 fall into this category.

The `junit`, `pylint`, and `rdjson` formats already included the `LintId` for syntax errors, so I similarly accepted those without further changes in 0dc0616cf7.

The other output formats did not fall back on the `LintId`, so I went through each of them in separate commits.
- azure: 0b625a1f6c
- github: a1488feea5
- grouped: 7cec0e02db
- sarif: 199af70fae
- json and json-lines: 1cdb26ce78

The `github` format seems a bit redundant because we print `invalid-syntax` twice, but that matches what we do for noqa codes too:

```diff
 ::error title=Ruff (F821),file=[TMP]/input.py,line=2,col=5,endLine=2,endColumn=6::input.py:2:5: F821 Undefined name `y`
-::error title=Ruff,file=[TMP]/input.py,line=3,col=1,endLine=3,endColumn=6::input.py:3:1: SyntaxError: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)
+::error title=Ruff (invalid-syntax),file=[TMP]/input.py,line=3,col=1,endLine=3,endColumn=6::input.py:3:1: invalid-syntax: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)
```

The `sarif` changes are slightly complicated by the `unique_rules` collection, which excludes syntax errors. I added the new enum to avoid needing to change the definition of `SarifRule::from` for `SecondaryCode`.

Initially it felt like a syntax error should also be a unique `SarifRule`, but it doesn't have most of the other fields like `linter`, `summary`, or `url`, and `name` and `code` would be the same.

ntBre commented:
I added a similar change for WASM as part of fixing that test failure. Instead of this:
<img width="813" height="256" alt="image" src="https://github.com/user-attachments/assets/44d6f9ad-75b0-4bf2-b301-9a8d69336f30" />

syntax errors now render as this in the playground:

<img width="817" height="258" alt="image" src="https://github.com/user-attachments/assets/e044226c-96de-43f5-8a3c-b8a6a1dc7202" />

MichaReiser reviewed (commented):
The snapshot changes look good to me. I'll review the rendering changes tomorrow

MichaReiser commented on python/ruff-ecosystem/ruff_ecosystem/check.py:
```diff
@@ -45,7 +45,7 @@
 
 CHECK_DIAGNOSTIC_LINE_RE = re.compile(
-    r"^(?P<diff>[+-])? ?(?P<location>.*): (?P<code>[A-Z]{1,4}[0-9]{3,4}|SyntaxError:)(?P<fixable> \[\*\])? (?P<message>.*)"
+    r"^(?P<diff>[+-])? ?(?P<location>.*): (?P<code>[A-Z]{1,4}[0-9]{3,4}|invalid-syntax:)(?P<fixable> \[\*\])? (?P<message>.*)"
```
We could probably change this to `[a-z\-]+` (I didn't test this pattern) to future prove (e.g. `io-error:`)

MichaReiser commented on crates/ruff_annotate_snippets/src/renderer/display_list.rs:
```diff
@@ -209,18 +214,51 @@ impl DisplaySet<'_> {
-                &format!("{}{}", annotation_type_str(&annotation.annotation_type), id),
-                *color,
-            );
+            // TODO(brent) All of this complicated checking of `hide_severity` should be reverted
```
Nice comment! This helped me a lot to understand what's going on. Thanks for writing it

MichaReiser commented on crates/ruff_annotate_snippets/src/renderer/display_list.rs:
```diff
@@ -920,6 +958,13 @@ pub(crate) enum DisplayAnnotationType {
 
+impl DisplayAnnotationType {
+    #[inline]
+    pub(crate) const fn is_none(&self) -> bool {
```
Can this be private?

MichaReiser reviewed (approved):
Thank you!

MichaReiser commented:
> It's still a bit of a shame about the missing cyan in the fixability marker, but I don't think there's a good way to fix that unless we add an is_fixable flag to the Annotation type in annotate-snippets

At this point it will be very hard to every use the upstream crate again because I don't think we can convince them to add this as a feature. But who knows, maybe the have something much more flexible. 

I think I'd be okay with adding it

MichaReiser commented on crates/ruff_annotate_snippets/src/renderer/display_list.rs:
```diff
@@ -253,6 +253,11 @@ impl DisplaySet<'_> {
             if let Some(id) = annotation.id {
                 if hide_severity {
                     buffer.append(line_offset, &format!("{id} "), *stylesheet.error());
+                    if annotation.is_fixable {
+                        buffer.append(line_offset, "[", stylesheet.none);
+                        buffer.append(line_offset, "*", stylesheet.help);
+                        buffer.append(line_offset, "] ", stylesheet.none);
+                    }
```
Do we need to gate this behind hide_severity or could we always render it after the severity/and id (unconditionally) if `is_fixable` is true
